DigitalLearningSolutions.Data/Services/CourseContentService.cs

```diff
@@ -23,21 +24,52 @@ public CourseContentService(IDbConnection connection, ILogger<CourseContentServi
             this.logger = logger;
         }
 
-        public CourseContent? GetCourseContent(int customisationId)
+        public CourseContent? GetCourseContent(int candidateId, int customisationId)
         {
-            return connection.QueryFirstOrDefault<CourseContent>(
-                @"SELECT Customisations.CustomisationID AS id,
+            CourseContent? courseContent = null;
+            return connection.Query<CourseContent, CourseSection, CourseContent>(
+                @"SELECT TOP 10000 Customisations.CustomisationID AS id,
                          Applications.ApplicationName,
-	                     Customisations.CustomisationName,
-	                     dbo.GetMinsForCustomisation(Customisations.CustomisationID) AS AverageDuration,
-	                     Centres.CentreName,
-	                     Centres.BannerText
+                         Customisations.CustomisationName,
+                         dbo.GetMinsForCustomisation(Customisations.CustomisationID) AS AverageDuration,
+                         Centres.CentreName,
+                         Centres.BannerText,
+                         Sections.SectionName,
+                         dbo.CheckCustomisationSectionHasLearning(Customisations.CustomisationID, Sections.SectionID) AS HasLearning,
+                         (CASE
+                            WHEN Progress.CandidateID IS NULL
+                                 OR dbo.CheckCustomisationSectionHasLearning(Customisations.CustomisationID, Sections.SectionID) = 0
+                            THEN 0
+                            ELSE CAST(SUM(aspProgress.TutStat) * 100 AS FLOAT) / (COUNT(Tutorials.TutorialID) * 2)
+                         END) AS PercentComplete
                   FROM Applications
                   INNER JOIN Customisations ON Applications.ApplicationID = Customisations.ApplicationID
+                  INNER JOIN Sections ON Sections.ApplicationID = Applications.ApplicationID
                   INNER JOIN Centres ON Customisations.CentreID = Centres.CentreID
-                  WHERE Customisations.CustomisationID = @customisationId;",
-                new { customisationId }
-            );
+                  LEFT JOIN Tutorials ON Sections.SectionID = Tutorials.SectionID
+                  LEFT JOIN Progress ON Customisations.CustomisationID = Progress.CustomisationID AND Progress.CandidateID = @candidateId
+                  LEFT JOIN aspProgress ON aspProgress.ProgressID = Progress.ProgressID AND aspProgress.TutorialID = Tutorials.TutorialID
+                  WHERE Customisations.CustomisationID = @customisationId
+                  GROUP BY

```

Comment:
Is the group by necessary? 🙂

Reply:
I think it might be, because of the aggregate functions `SUM` and `COUNT` on line 43. There might be another way of doing this though, maybe using subqueries? 🙂

Reply:
OKay, fair enough 🙂